circuit-types/src/traits.rs

```diff
@@ -1078,87 +554,75 @@ pub trait SingleProverCircuit {
-    const BP_GENS_CAPACITY: usize;
-
     /// Apply the constraints of the circuit to a given constraint system
-    fn apply_constraints<CS: RandomizableConstraintSystem>(
-        witness_var: <Self::Witness as CircuitBaseType>::VarType<Variable>,
-        statement_var: <Self::Statement as CircuitBaseType>::VarType<Variable>,
-        cs: &mut CS,
-    ) -> Result<(), R1CSError>;
+    fn apply_constraints(
+        witness_var: <Self::Witness as CircuitBaseType>::VarType,
+        statement_var: <Self::Statement as CircuitBaseType>::VarType,
+        cs: &mut PlonkCircuit,
+    ) -> Result<(), PlonkError>;
 
     /// Generate a proof of the statement represented by the circuit
     ///
     /// Returns both the commitment to the inputs, as well as the proof itself
     fn prove(
         witness: Self::Witness,
         statement: Self::Statement,
-        bp_gens: &BulletproofGens,
-        mut prover: Prover,
-    ) -> Result<
-        (
-            <Self::Witness as CircuitBaseType>::CommitmentType,
-            R1CSProof,
-        ),
-        ProverError,
-    > {
-        // Commit to the witness and statement
-        let mut rng = thread_rng();
-        let (witness_var, witness_comm) = witness.commit_witness(&mut rng, &mut prover);
-        let statement_var = statement.commit_public(&mut prover);
+        pk: &ProvingKey<SystemCurve>,
+        mut circuit: PlonkCircuit,
+    ) -> Result<Proof<SystemCurve>, ProverError> {
+        // Allocate the witness and statement in the constraint system
+        let witness_var = witness.create_witness(&mut circuit);
+        let statement_var = statement.create_public_var(&mut circuit);
 
         // Apply the constraints
-        Self::apply_constraints(witness_var, statement_var, &mut prover)
-            .map_err(ProverError::R1CS)?;
+        Self::apply_constraints(witness_var, statement_var, &mut circuit)
+            .map_err(ProverError::Plonk)?;
 
         // Generate the proof
-        let proof = prover.prove(bp_gens).map_err(ProverError::R1CS)?;
-
-        Ok((witness_comm, proof))
+        let mut rng = thread_rng();
+        PlonkKzgSnark::prove::<_, _, SolidityTranscript>(
+            &mut rng, &circuit, pk, None, // extra_init_msg
+        )
+        .map_err(ProverError::Plonk)
     }
 
     /// Verify a proof of the statement represented by the circuit
     ///
-    /// The verifier has access to the statement variables, but only hiding (and
-    /// binding) commitments to the witness variables
+    /// The verifier has access to the statement variables, but not the witness
     fn verify(
-        witness_commitment: <Self::Witness as CircuitBaseType>::CommitmentType,
         statement: Self::Statement,
-        proof: R1CSProof,
-        bp_gens: &BulletproofGens,
-        mut verifier: Verifier,
+        proof: Proof<SystemCurve>,
+        vk: &VerifyingKey<SystemCurve>,
     ) -> Result<(), VerifierError> {
         // Commit to the witness and statement
```

Comment:
Nit: we shouldn't mention the witness here, and we shouldn't say that we're "committing" to the statement, perhaps just "Serialize the statement into scalars"